core/services/relay/evm/chain_reader.go

```diff
@@ -23,132 +28,223 @@ type ChainReaderService interface {
-		return fmt.Errorf("Failed to initialize ChainReader: %w", err)
+func (cr *chainReader) GetLatestValue(ctx context.Context, contractName, method string, params any, returnVal any) error {
+	ae, err := cr.bindings.getBinding(contractName, method, false)
+	if err != nil {
+		return err
 	}
-	return nil
-}
-
-func (cr *chainReader) Close() error { return nil }
 
-func (cr *chainReader) Ready() error { return nil }
+	if ae.evt == nil {
+		return cr.getLatestValueFromContract(ctx, contractName, method, params, returnVal)
+	}
 
-func (cr *chainReader) HealthReport() map[string]error {
-	return map[string]error{cr.Name(): nil}
+	return cr.getLatestValueFromLogPoller(ctx, contractName, method, *ae.evt, returnVal)
 }
 
-func (cr *chainReader) GetLatestValue(ctx context.Context, bc commontypes.BoundContract, method string, params any, returnVal any) error {
-	return commontypes.UnimplementedError("Unimplemented method GetLatestValue called")
+func (cr *chainReader) getLatestValueFromLogPoller(ctx context.Context, contractName, method string, hash common.Hash, returnVal any) error {
+	ae, err := cr.bindings.getBinding(contractName, method, false)
+	if err != nil {
+		return err
+	}
+
+	log, err := cr.lp.LatestLogByEventSigWithConfs(hash, ae.addr, logpoller.Finalized)
+	if err != nil {
+		if strings.Contains(err.Error(), "not found") {
+			return fmt.Errorf("%w: %w", commontypes.ErrNotFound, err)
+		}
+		return fmt.Errorf("%w: %w", commontypes.ErrInternal, err)
+	}
+	return cr.codec.Decode(ctx, log.Data, returnVal, wrapItemType(contractName, method, false))
 }
 
-func validateChainReaderConfig(cfg types.ChainReaderConfig) error {
-	if len(cfg.ChainContractReaders) == 0 {
-		return fmt.Errorf("%w: no contract readers defined", commontypes.ErrInvalidConfig)
+func (cr *chainReader) getLatestValueFromContract(ctx context.Context, contractName, method string, params any, returnVal any) error {
+	data, err := cr.codec.Encode(ctx, params, wrapItemType(contractName, method, true))
+	if err != nil {
+		return err
 	}
 
-	for contractName, chainContractReader := range cfg.ChainContractReaders {
-		abi, err := abi.JSON(strings.NewReader(chainContractReader.ContractABI))
-		if err != nil {
-			return fmt.Errorf("invalid abi: %w", err)
-		}
+	ae, err := cr.bindings.getBinding(contractName, method, true)
+	if err != nil {
+		return err
+	}
+	callMsg := ethereum.CallMsg{
+		To:   &ae.addr,
+		From: ae.addr,
+		Data: data,
+	}
 
-		for chainReadingDefinitionName, chainReaderDefinition := range chainContractReader.ChainReaderDefinitions {
-			switch chainReaderDefinition.ReadType {
-			case types.Method:
-				err = validateMethods(abi, chainReaderDefinition)
-			case types.Event:
-				err = validateEvents(abi, chainReaderDefinition)
-			default:
-				return fmt.Errorf("%w: invalid chainreading definition read type: %d for contract: %q", commontypes.ErrInvalidConfig, chainReaderDefinition.ReadType, contractName)
+	output, err := cr.client.CallContract(ctx, callMsg, nil)
```

Comment:
I think the last param here is block number, where nil stands for latest pending block.  We should check `ae.pending`, if that's false we can use lastest finalized block instead.